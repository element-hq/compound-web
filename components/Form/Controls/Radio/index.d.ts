export { RadioControl, RadioInput } from "./Radio";
//# sourceMappingURL=index.d.ts.map