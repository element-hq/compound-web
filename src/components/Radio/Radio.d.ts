import React, { PropsWithChildren } from "react";
type RadioProps = {
    kind?: "primary" | "critical";
    className?: string;
} & React.ComponentPropsWithoutRef<"input">;
/**
 * Radio form control
 */
export declare const Radio: ({ kind, className, ...props }: PropsWithChildren<RadioProps>) => React.JSX.Element;
export {};
