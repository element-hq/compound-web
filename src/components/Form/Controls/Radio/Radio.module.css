/*
Copyright 2025 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.
Copyright 2023 New Vector Ltd

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

.container {
  --size: 20px;

  display: grid;
  inline-size: var(--size);
  block-size: var(--size);
}

.input,
.ui {
  box-sizing: border-box;
  grid-area: 1/1;
  inline-size: var(--size);
  block-size: var(--size);
}

.input {
  opacity: 0;
  margin: 0;
  cursor: pointer;
}

.ui {
  pointer-events: none;
  border-radius: 50%;
  border: 1px solid;
  border-color: var(--cpd-color-border-interactive-primary);

  /* To align the ::after pseudo-element to the center of the radio button */
  display: flex;
  align-items: center;
  justify-content: center;
}

.ui::after {
  content: "";
  inline-size: 6px;
  block-size: 6px;
  border-radius: 50%;
  background: transparent;
}

.input:checked + .ui {
  background-color: var(--cpd-color-bg-accent-rest);
  border-color: var(--cpd-color-bg-accent-rest);
}

.input:checked + .ui::after {
  background: var(--cpd-color-icon-on-solid-primary);
}

.input:focus-visible + .ui {
  outline: 2px solid var(--cpd-color-border-focused);
  outline-offset: 1px;
}

.input[readonly] {
  pointer-events: none;
}

.input[readonly] + .ui {
  border-color: var(--cpd-color-border-interactive-secondary);
  background: var(--cpd-color-bg-subtle-secondary);
}

.input[disabled] + .ui {
  border-color: var(--cpd-color-border-disabled);
  background: var(--cpd-color-bg-canvas-disabled);
}

.input[disabled]:checked + .ui {
  border-color: var(--cpd-color-bg-action-primary-disabled);
  background: var(--cpd-color-bg-action-primary-disabled);
}

.input[readonly]:checked + .ui::after {
  background-color: var(--cpd-color-icon-secondary);
}

@media (hover) {
  .input:not([disabled], [readonly], :checked):hover + .ui {
    border-color: var(--cpd-color-bg-accent-hovered);

    /** TODO: have the shadow in the design tokens */
    box-shadow: 0 1.2px 2.4px 0 rgb(0 0 0 / 15%);
  }

  .input:not([disabled], [readonly], :checked):hover + .ui::after {
    background: var(--cpd-color-icon-quaternary);
  }

  .input:not([disabled], [readonly]):checked:hover + .ui {
    border-color: var(--cpd-color-bg-accent-hovered);
    background: var(--cpd-color-bg-accent-hovered);
  }

  .input[data-invalid]:not([disabled], [readonly]):checked:hover + .ui {
    border-color: var(--cpd-color-bg-critical-hovered);
    background: var(--cpd-color-bg-critical-hovered);
  }
}

.input[data-invalid]:not([disabled], [readonly], :checked) + .ui {
  border-color: var(--cpd-color-border-critical-primary);
}

.input[data-invalid]:not([disabled], [readonly]):checked + .ui {
  background-color: var(--cpd-color-bg-critical-primary);
  border-color: var(--cpd-color-bg-critical-primary);
}
