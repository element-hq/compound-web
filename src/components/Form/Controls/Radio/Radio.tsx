/*
Copyright 2025 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.
Copyright 2023 New Vector Ltd

SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
Please see LICENSE files in the repository root for full details.
*/

import classnames from "classnames";
import React, {
  type ComponentProps,
  type ComponentRef,
  forwardRef,
  type PropsWithChildren,
} from "react";
import { Control } from "@radix-ui/react-form";

import styles from "./Radio.module.css";

type RadioProps = {
  /**
   * The CSS class name.
   */
  className?: string;
} & Omit<React.ComponentPropsWithoutRef<"input">, "type">;

/**
 * A radio component.
 */
export const RadioInput = forwardRef<
  HTMLInputElement,
  PropsWithChildren<RadioProps>
>(function Radio({ className, ...props }, ref) {
  const classes = classnames(styles.container, className);
  return (
    <div className={classes}>
      <input ref={ref} {...props} className={styles.input} type="radio" />
      <div className={styles.ui} />
    </div>
  );
});

export const RadioControl = forwardRef<
  ComponentRef<typeof RadioInput>,
  ComponentProps<typeof RadioInput>
>(function RadioControl(props, ref) {
  return (
    <Control asChild>
      <RadioInput ref={ref} {...props} />
    </Control>
  );
});
