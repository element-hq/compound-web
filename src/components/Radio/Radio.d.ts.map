{"version":3,"file":"Radio.d.ts","sourceRoot":"","sources":["../../../../src/components/Radio/Radio.tsx"],"names":[],"mappings":"AAiBA,OAAO,KAAK,EAAE,EAAE,iBAAiB,EAAE,MAAM,OAAO,CAAC;AAGjD,KAAK,UAAU,GAAG;IAChB,IAAI,CAAC,EAAE,SAAS,GAAG,UAAU,CAAC;IAC9B,SAAS,CAAC,EAAE,MAAM,CAAC;IACnB,WAAW,CAAC,EAAE,CAAC,CAAC,EAAE,KAAK,CAAC,UAAU,CAAC,gBAAgB,EAAE,UAAU,CAAC,KAAK,IAAI,CAAC;CAC3E,GAAG,KAAK,CAAC,wBAAwB,CAAC,OAAO,CAAC,CAAC;AAE5C;;GAEG;AACH,eAAO,MAAM,KAAK,+CAKf,kBAAkB,UAAU,CAAC,sBAiB/B,CAAC"}