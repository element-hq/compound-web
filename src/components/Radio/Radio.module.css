/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.radio {
  --size: 24px;

  display: grid;
  width: var(--size);
  height: var(--size);
}

.radio > * {
  grid-area: 1/1;
  width: var(--size);
  height: var(--size);
}

.radio > input[type="radio"] {
  opacity: 0;
  margin: 0;
  cursor: pointer;
}

.radio-ui {
  pointer-events: none;
  width: 24px;
  box-sizing: border-box;
  aspect-ratio: 1 / 1;
  border-radius: 50%;
  border: 1px solid;
  display: flex;
  align-items: center;
  justify-content: center;
}

.radio-ui::after {
  content: "";
  display: block;
  width: 10px;
  height: 10px;
  border-radius: 50%;
  background: transparent;
}

:checked + .radio-ui::after {
  background: var(--cpd-color-icon-on-solid-primary);
}

:hover + .radio-ui::after {
  background: var(--cpd-color-icon-quaternary);
}

:checked:hover + .radio-ui::after {
  background: var(--cpd-color-icon-on-solid-primary);
}

[data-kind="primary"] .radio-ui {
  border-color: var(--cpd-color-border-interactive-primary);
}

[data-kind="critical"] .radio-ui {
  border-color: var(--cpd-color-border-critical-primary);
}

[type="radio"]:focus + .radio-ui {
  outline: 2px solid var(--cpd-color-blue-900);
  outline-offset: 1px;
}

.radio[data-kind="primary"] :checked + .radio-ui {
  background: var(--cpd-color-bg-action-primary-rest);
  border-color: var(--cpd-color-bg-action-primary-rest);
}

.radio[data-kind="critical"] :checked + .radio-ui {
  background: var(--cpd-color-bg-critical-primary);
  border-color: var(--cpd-color-bg-critical-primary);
}

.radio[data-kind="primary"] [type="radio"]:hover + .radio-ui {
  color: var(--cpd-color-icon-quaternary);
  border-color: var(--cpd-color-border-interactive-hovered);
}

.radio[data-kind="critical"] [type="radio"]:hover + .radio-ui {
  color: var(--cpd-color-icon-quaternary);
  border-color: var(--cpd-color-border-critical-hovered);
}

.radio[data-kind] [type="radio"][disabled] + .radio-ui {
  border-color: var(--cpd-color-border-disabled);
  background: var(--cpd-border-disabled);
  color: var(--color-text-disabled);
}

.radio[data-kind] [type="radio"][disabled]:checked + .radio-ui {
  background: var(--cpd-color-bg-action-primary-disabled);
}

.radio[data-kind="primary"] [type="radio"]:checked:hover + .radio-ui {
  background: var(--cpd-color-bg-action-primary-hovered);
}

.radio[data-kind="critical"] [type="radio"]:checked:hover + .radio-ui {
  background: var(--cpd-color-bg-critical-hovered);
}
