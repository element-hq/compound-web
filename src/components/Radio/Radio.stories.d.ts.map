{"version":3,"file":"Radio.stories.d.ts","sourceRoot":"","sources":["../../../../src/components/Radio/Radio.stories.tsx"],"names":[],"mappings":"AAgBA,OAAO,KAAK,MAAM,OAAO,CAAC;;;;;;;;AAK1B,wBAKiC;AAMjC,eAAO,MAAM,OAAO;;;;;;EAAoB,CAAC;AAIzC,eAAO,MAAM,QAAQ;;;;;;EAAoB,CAAC"}