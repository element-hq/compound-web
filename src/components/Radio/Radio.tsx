/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import classnames from "classnames";
import React, { PropsWithChildren } from "react";
import styles from "./Radio.module.css";

type RadioProps = {
  kind?: "primary" | "critical";
  className?: string;
} & React.ComponentPropsWithoutRef<"input">;

/**
 * Radio form control
 */
export const Radio = ({
  kind = "primary",
  className,
  ...props
}: PropsWithChildren<RadioProps>) => {
  const classes = classnames(styles.radio, className);
  return (
    <div className={classes} data-kind={kind}>
      <input {...props} type="radio" />
      <div className={styles["radio-ui"]} />
    </div>
  );
};
