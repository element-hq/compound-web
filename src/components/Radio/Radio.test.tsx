/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { describe, it, expect } from "vitest";
import { getByRole, render } from "@testing-library/react";
import React from "react";

import { Radio } from "./Radio";
import userEvent from "@testing-library/user-event";

describe("Radio", () => {
  it("renders", () => {
    const { asFragment } = render(<Radio />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should not have focus after being clicked", async () => {
    const { container } = render(<Radio />);
    const radio = getByRole(container, "radio");
    await userEvent.click(radio);
    expect(radio).not.toHaveFocus();
  });
});
