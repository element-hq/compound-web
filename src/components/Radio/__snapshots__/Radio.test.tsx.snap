// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Radio > renders 1`] = `
<DocumentFragment>
  <div
    class="_radio_4e7d3c"
    data-kind="primary"
  >
    <input
      type="radio"
    />
    <div
      class="_radio-ui_4e7d3c"
    />
  </div>
</DocumentFragment>
`;
